### ISSUE TITLE
P-GAM fitting becomes extremely slow with neuron-to-neuron coupling filters added in large population models

### ISSUE THREAD

**Issue by HeliyaShakeri** [BalzaniEdoardo/PGAM#5]
Hi, thanks for the package.

We are using P-GAM for spike-train modeling with 5 temporal task variables (whisker, lick, reward and auditory events). 

We then add neuron-to-neuron coupling filters:
- within-area coupling: 4 bins
- across-area coupling: 60 bins
- coupling only for neurons from the same session_id which gives 500-1000 possible coupled neurons per target neuron.

Our data are roughly:
- 400 trials
- ~4 s per trial
- 10 ms bins

The problem is that once we included the coupling covariates, fitting a single target neuron took several hours and still did not fitted. 

We wanted to ask:
- Is this expected runtime for large coupling models in P-GAM, or does it suggest a problem in our implementation?
- Are there practical recommendations for this number of coupled neurons / coupling terms?

**Comment by BalzaniEdoardo** [comment]
Hi, thanks for this issue. There are two things that may be happening: 

1. The fit is not converging because it is oscillating between two equivalent solutions. Usually, the fit terminates in tens of iterations, so if you are observing a large number of iterations that could be the issue. In my experience, it is very rare but can happen, and it should be happening only for a handful of neuron unless there is a poorly conditioned model design. This may happen if some neurons are not spiking or if there are neurons that are double counted or similar.

2. Each iteration is very slow, and even 10 iterations take many minutes. This suggests that the number of features is too large for the current implementation. Internally, there is an SVD decomposition of a matrix that contains the counts convolved with the basis. If this matrix has a very large number of columns, the svd will be super slow. This is a limitation of the current implementation. 

If your case is 1, then some pre-processing and filtering of the units (like keeping neurons with firing rates above a threshold) should help.

If it is 2, we are working on different approaches:
a) a gpu accelerated version of the computation, this would be perform better if you have a JAX-cuda installed.
b) we are thinking of a version of the PGAM that fits very differently and would work perfectly in your case: a Gaussian process based inference, that can assume all trials to be independent. The fit in your case should be very efficient.

@bagibence, if we have a gpu accelerated version of the code that can be tested, can you point that out to @HeliyaShakeri?
@HeliyaShakeri are your trials of the same duration?
@pgree do you think the GP-approach is feasible for the problem described above?

**Comment by HeliyaShakeri** [comment]
Thanks, this is very helpful.

Yes, all of our trials are the same duration: 4 s.

From what we have seen so far, this seems more likely to be case 2 than case 1, because the model without coupling terms fits normally, and the extreme slowdown appears only after adding the neuron-to-neuron coupling covariates.

The GPU/JAX version would definitely be interesting to test if it is available.

Many thanks again.

**Comment by BalzaniEdoardo** [comment]
@colecitrenbaum this is a type of dataset in which the GP approach we discussed would scale better then the current implementation.  The data structure is very similar to the simulation I have in the notebook of this repo, but with many more covariates, all 1D.

**Comment by BalzaniEdoardo** [comment]
@HeliyaShakeri I have a practical suggestion for you, as we are waiting for the jax version and/or the other algorithmic improvements. Since most neuron pairs are likely not functionally connected, you could first run a sparse GLM (e.g., Elastic Net) as a screening step. Rather than fixing a regularization strength a priori, I’d recommend either cross-validation with a bias toward lower regularization (to favor inclusion), or stability selection to identify consistently selected predictors. Then, fit the GAM only on the subset of neuron pairs that survive this screening.

If you want a GPU accelerated version of the GLM with all the regularization schemes cited above implemented, look at this package:
https://nemos.readthedocs.io/en/latest/

```python
import nemos as nmo

# example elastic net
model = nmo.glm.GLM(regularizer="ElasticNet", regularizer_strength=0.001)

# simulate counts is (n_samples, n_neurons)
n_samples, n_neurons = 500, 3
counts = np.random.poisson(size=(500, 3))

# define a BSpline (other basis option are available but this matches the GAM package)
 # 4 basis, window size = 100 bins
X = nmo.basis.BSplineConv(
    n_basis_funcs=4, window_size=100
).compute_features(counts)


# fit first neuron
model.fit(X, counts[:, 0])
```

There is also a population GLM that fits the total log likelihood, however, for such a large population I recommend to fit neruons individually, on a GPU should be doable as long as X fit in memory.

**Comment by bagibence** [comment]
A very-very early JAX version is now available at https://github.com/bagibence/pgam_jax

**Comment by HeliyaShakeri** [comment]
Hello, 

Thank you so much for the comments. I used the JAX version and it worked. 

Many thanks.

**Comment by BalzaniEdoardo** [comment]
@HeliyaShakeri thanks for the update, great news! 
We are actively working on that code and we keep improving it. There will be breaking changes most likely, so that could be annoying at this stage, but it is worth keeping an eye out for updates
